compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIETileForMatmulLikeInnerLoop.cpp

```diff
@@ -0,0 +1,168 @@
+      auto numDimsFromFirstNonSingleton1 =
+          getNumDimsFromFirstNonSingleton(genericOp->getOperand(1));
+      auto numDimsFromFirstNonSingleton2 =
+          getNumDimsFromFirstNonSingleton(genericOp->getOperand(2));
+
+      if (numDimsFromFirstNonSingleton0 < 3 &&
+          numDimsFromFirstNonSingleton1 < 3 &&
+          numDimsFromFirstNonSingleton2 < 3) {
+        return rewriter.notifyMatchFailure(
+            genericOp,
+            llvm::formatv(
+                "The number of non-singleton trailing dimensions is less than "
+                "3 for all operands ({0}, {1}, {2}), so there are no outer "
+                "dimensions to tile.",
+                numDimsFromFirstNonSingleton0, numDimsFromFirstNonSingleton1,
+                numDimsFromFirstNonSingleton2));
+      }
+    }
+
+    assert(iteratorTypes.size() >= 3 && "expected at least 3 iterators here");
+
+    auto hasMatmulIterators = [&]() {
+      return linalg::isParallelIterator(
+                 iteratorTypes[iteratorTypes.size() - 3]) &&
+             linalg::isParallelIterator(
+                 iteratorTypes[iteratorTypes.size() - 2]) &&
+             linalg::isReductionIterator(
+                 iteratorTypes[iteratorTypes.size() - 1]);
+    }();
+
+    if (!hasMatmulIterators) {
+      return rewriter.notifyMatchFailure(
+          genericOp,
+          "The final 3 dimensions do not match matmul iterators (should be "
+          "'parallel parallel reduction').");
+    }
+
+    // Check that the mapped dimensions correspond to matmul.
+    // Specifically the final dimensions of the maps are
+    // operand 0 : (i, j, k) -> (i, k)
+    // operand 1 : (i, j, k) -> (k, j)
+    // result    : (i, j, k) -> (i, j)
+
+    auto indexingMaps = genericOp.getIndexingMaps();
+    assert(indexingMaps.size() == 3 && "expected 3 indexing maps here");
+
+    AffineMap map0 = cast<AffineMapAttr>(indexingMaps[0]).getValue();
+    AffineMap map1 = cast<AffineMapAttr>(indexingMaps[1]).getValue();
+    AffineMap map2 = cast<AffineMapAttr>(indexingMaps[2]).getValue();
+
+    auto nResults0 = map0.getNumResults();
+    auto nResults1 = map1.getNumResults();
+    auto nResults2 = map2.getNumResults();
+
+    if (map0.getResult(nResults0 - 2) != map2.getResult(nResults2 - 2))
+      return rewriter.notifyMatchFailure(
+          genericOp,
+          "unable to locate 'M' in affine maps [(M,K) x (K,N) -> (M,N)]");
+    if (map1.getResult(nResults1 - 1) != map2.getResult(nResults2 - 1))
+      return rewriter.notifyMatchFailure(
+          genericOp,
+          "unable to locate 'N' in affine maps [(M,K) x (K,N) -> (M,N)]");
+    if (map0.getResult(nResults0 - 1) != map1.getResult(nResults1 - 2))
+      return rewriter.notifyMatchFailure(
+          genericOp,
+          "unable to locate 'K' in affine maps [(M,K) x (K,N) -> (M,N)]");
+
+    // Ran the gauntlet, now perform the tiling of the outer-dimensions.
```

Comment:
The file name `AMDAIETileForMatmulLikeInnerLoop` is a little confusing and seems contradicting this comment?

Reply:
Do you mean the pass name `--iree-amdaie-tile-for-matmul-like-inner-loop` could be more descriptive? 


The pass tiles a linalg.generic so that it becomes an scf.for loop with a simple matmul inside. I've gone with matmul "like" in the pass name because I'm not checking that the reduction is actually a multiply accumulate. I could add this check, but it seems like unneccessary complexity. 

Some options:

--iree-amdaie-decompose-to-loop-with-simple-matmul 
--iree-amdaie-tile-outer-dims-of-matmul-like-generic
--iree-amdaie-tile-for-matmul-like-inner-loop
--iree-amdaie-insert-loops-for-vectorization

or combinations of the above :-)

Reply:
> The file name AMDAIETileForMatmulLikeInnerLoop is a little confusing and seems contradicting this comment?

The tiling is trivial: Use tile size of 1 for all but dimensions except the 3 inner-most dimensions (which don't get tiled, i.e. get tiles with '0'). So the outer dimensions get tiled.

Reply:
I think `--iree-amdaie-insert-loops-for-vectorization` maybe a better choice.